File: python/algo/leetcode3/algo2999.py
from functools import lru_cache
class Solution:
    
    def numberOfPowerfulInt(self, start: int, finish: int, limit: int, s: str) -> int:

        sf = str(finish)
        ss = str(start).rjust(len(sf), '0')
        diff = len(sf) - len(s)

        @lru_cache(maxsize=None)
        def f(i:int, low_limit:bool, up_limit:bool):
            if i == len(sf):
                return 1

            lo = int(ss[i]) if low_limit else 0
            hi = int(sf[i]) if up_limit else 9

            ans = 0
            if i < diff:
                for d in range(lo, min(hi+1,limit+1)):
                    ans += f(i+1, low_limit and d==lo, up_limit and d==hi)
            else:
                x = int(s[i-diff])
                if lo <= x <= min(hi,limit):
                    ans += f(i+1, low_limit and x==lo, up_limit and x==hi)
            return ans
        return f(0, True, True)
    

if __name__ == "__main__":
    sol = Solution()
    print(sol.numberOfPowerfulInt(1, 6000, 4, "124"))

File: cplus/leetcode/algo2999.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    long long numberOfPowerfulInt(long long start, long long finish, int limit, string s)
    {
        string sf = to_string(finish);
        string ss = to_string(start);
        ss.insert(ss.begin(), sf.size() - ss.size(), '0');
        int diff = sf.size() - s.size();

        vector<long long> memo(sf.size(), -1);

        function<long long(int, bool, bool)> f = [&](int i, bool limit_low, bool limit_up)
        {
            if (i == sf.size())
            {
                return (long long)1;
            }

            if (!limit_low && !limit_up && memo[i] != -1)
                return memo[i];


            int lo = (limit_low) ? stoi(ss.substr(i, 1)) : 0;
            int hi = (limit_up) ? stoi(sf.substr(i, 1)) : 9;

            long long ans = 0;

            if (i < diff)
            {
                for (int k = lo; k <= hi && k <= limit; k++)
                    ans += f(i + 1, limit_low && k == lo, limit_up && k == hi);
            }
            else
            {
                int x = stoi(s.substr(i - diff, 1));
                if (x >= lo && x <= hi && x <= limit)
                    ans += f(i + 1, limit_low && x == lo, limit_up && x == hi);
            }
            if (!limit_low && !limit_up) {
                memo[i] = ans;
            }
            return ans;
        };

        return f(0, true, true);
    }
};

int main()
{
    Solution solution;

    cout << solution.numberOfPowerfulInt(15, 215, 6, "10") << endl;
}